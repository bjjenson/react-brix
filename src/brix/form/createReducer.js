import { fromJS, Map } from 'immutable'
import { useReducer, useMemo } from 'react'
import { fieldReducer, initState, actions } from './fieldReducer'
import { defaultTextValue } from './fields/useTextField'
import { defaultSelectValue } from './fields/useSelectField'
import { defaultNumberValue } from './fields/useNumberField'
import { defaultBooleanValue } from './fields/useBooleanField'

export const createReducer = ({ fields, initialValues = Map() }) => {
  const derivedState = getInitialState(fields, initialValues)
  const [state, dispatch] = useReducer(fieldReducer, derivedState, initState)

  useMemo(() => {
    dispatch(actions.reset(derivedState))
    return initialValues
  }, [initialValues])

  return [state, dispatch]
}

/**
 * @param fields {Array<import("../form").IFormFieldArgs>}
 */
const getInitialState = (fields, initialValues) => {
  const fieldMap = fields.reduce((acc, field) => {
    return acc.set(field.name, generateDefaultFieldState(field, initialValues))
  }, Map())

  return fieldMap
}

export const generateDefaultFieldState = (field, initialValues) => {
  if (!field.name) throw new Error('name is required on field')

  const value = getInitialValue(field.value, initialValues.get(field.name), field.type)


  return fromJS({
    initial: {
      type: field.type || 'text',
      value,
      optional: field.optional || false,
      label: field.label,
    },
    current: {
      helperText: field.helperText || '',
      error: false,
      pristine: true,
      touched: false,
      value,
    },
  }).setIn(['initial', 'field'], field)
}

const defaultValues = {
  text: defaultTextValue,
  select: defaultSelectValue,
  number: defaultNumberValue,
  boolean: defaultBooleanValue,
}

const getInitialValue = (fieldValue, initialValue, type = 'text') => {
  if (initialValue) {
    return initialValue
  }
  if (fieldValue) {
    return fieldValue
  }
  return defaultValues[type]
}
