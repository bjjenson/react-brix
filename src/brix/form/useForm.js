import { Map } from 'immutable'
import { Form } from './Form'
import { createReducer, generateDefaultFieldState } from './createReducer'
import { actions } from './fieldReducer'
import {
  useBooleanField,
  useNumberField,
  useSelectField,
  useTextField,
} from './fields'

/**
 * @param  param0 { import("../form").IFormProps }
 */
export const useForm = ({ fields, submit, validate, options = {}, initialValues = Map() }) => {
  const [state, dispatch] = createReducer({ fields, initialValues, options })

  const fieldData = state.keySeq().reduce((acc, fieldName) => {
    const fieldState = state.get(fieldName)
      .set('getAllValues', () => getFieldValues(fieldData))

    const fieldType = state.getIn([fieldName, 'initial', 'type'])
    const field = state.getIn([fieldName, 'initial', 'field'])
    switch (fieldType) {
      case 'select':
        acc[fieldName] = useSelectField(fieldState, dispatch, field)
        break
      case 'boolean':
        acc[fieldName] = useBooleanField(fieldState, dispatch, field)
        break
      case 'number':
        acc[fieldName] = useNumberField(fieldState, dispatch, field)
        break
      case 'text':
      default:
        acc[fieldName] = useTextField(fieldState, dispatch, field)
        break
    }
    return acc
  }, {})

  const addField = field => {
    const fieldState = generateDefaultFieldState(field, initialValues)
    dispatch(actions.insertField(field.name, fieldState))
  }

  const removeField = fieldName => {
    dispatch(actions.removeField(fieldName))
  }

  const tryValidateForm = () => {
    if (validate) {
      const values = getFieldValues(fieldData)
      return validate(values)
    }

    return {}
  }

  const trySubmitForm = () => {
    let isFormValid = true
    const formResults = tryValidateForm()
    const values = Object.entries(fieldData).reduce((acc, [key, v]) => {
      if (formResults[key]) {
        v.setValidationResult({ error: true, helperText: formResults[key] })
        isFormValid = false
      }

      const isFieldValid = v.validate()
      isFormValid = isFormValid && isFieldValid

      return acc.set(key, v.props.value)
    }, initialValues)
    if (isFormValid) {
      submit(values)
    }
  }

  const setValue = (fieldName, value) => {
    fieldData[fieldName].setValue(value)
  }

  return {
    ...getFieldProps(fieldData),
    setValue,
    submit: trySubmitForm,
    Form,
    addField,
    removeField,
  }
}

const getFieldValues = (fieldData) => {
  return Object.entries(fieldData).reduce((acc, [k, v]) => {
    acc[k] = v.props.value
    return acc
  }, {})
}

const getFieldProps = (fieldData) => {
  return Object.entries(fieldData).reduce((acc, [k, v]) => {
    acc[k] = v.props
    return acc
  }, {})
}
