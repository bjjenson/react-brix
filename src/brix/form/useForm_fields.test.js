import { fromJS } from 'immutable'
import { useForm } from './useForm'
import {
  useBooleanField,
  useNumberField,
  useSelectField,
  useTextField,
} from './fields'
import { createReducer } from './createReducer'

jest.mock('./fields')
jest.mock('./createReducer')

const setValue = jest.fn()

let fieldProps
beforeEach(() => {
  fieldProps = {
    props: {},
    setValue,
    validate: jest.fn(),
  }

  const createState = type => fromJS({
    fieldName: {
      initial: {
        type,
      },
      current: {
        value: 'value of fieldName',
      },
    },
  })

  createReducer.mockImplementation(({ fields }) => {
    return [createState(fields[0].type || 'text'), 'dispatch']
  })
  useBooleanField.mockReturnValue(fieldProps)
  useNumberField.mockReturnValue(fieldProps)
  useSelectField.mockReturnValue(fieldProps)
  useTextField.mockReturnValue(fieldProps)
})

describe('text', () => {
  test('defaults to text', () => {
    useForm({
      fields: [
        { name: 'fieldName', label: 'the label' },
      ],
    })

    expect(useTextField.mock.calls[0]).toMatchSnapshot()
  })

  test('uses text', () => {
    useForm({
      fields: [
        { name: 'fieldName', type: 'text', label: 'the label' },
      ],
    })

    expect(useTextField.mock.calls[0]).toMatchSnapshot()
  })

  test('setValue updates field value', () => {
    const form = useForm({
      fields: [
        { name: 'fieldName', type: 'text', label: 'the label' },
      ],
    })
    form.setValue('fieldName', 'new value')
    expect(setValue.mock.calls[0]).toMatchSnapshot()
  })
})

describe('number', () => {
  test('uses text', () => {
    useForm({
      fields: [
        { name: 'fieldName', type: 'number', label: 'the label' },
      ],
    })

    expect(useNumberField.mock.calls[0]).toMatchSnapshot()
  })

})

describe('select', () => {
  test('uses select', () => {
    useForm({
      fields: [
        { name: 'fieldName', type: 'select', label: 'the label' },
      ],
    })

    expect(useSelectField.mock.calls[0]).toMatchSnapshot()
  })

})

describe('boolean', () => {
  test('uses boolean', () => {
    useForm({
      fields: [
        { name: 'fieldName', type: 'boolean', label: 'the label' },
      ],
    })

    expect(useBooleanField.mock.calls[0]).toMatchSnapshot()
  })

})
