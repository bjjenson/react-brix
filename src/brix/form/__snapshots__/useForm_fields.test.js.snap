// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`boolean passes initialValue 1`] = `
Array [
  Object {
    "label": "the label",
    "name": "fieldName",
    "type": "boolean",
  },
  "i am initial",
]
`;

exports[`boolean uses boolean 1`] = `
Array [
  Object {
    "label": "the label",
    "name": "fieldName",
    "type": "boolean",
  },
  undefined,
]
`;

exports[`number passes initialValue 1`] = `
Array [
  Object {
    "label": "the label",
    "name": "fieldName",
    "type": "number",
  },
  "i am initial",
]
`;

exports[`number uses text 1`] = `
Array [
  Object {
    "label": "the label",
    "name": "fieldName",
    "type": "number",
  },
  undefined,
]
`;

exports[`select passes initialValue 1`] = `
Array [
  Object {
    "label": "the label",
    "name": "fieldName",
    "type": "select",
  },
  "i am initial",
]
`;

exports[`select uses select 1`] = `
Array [
  Object {
    "label": "the label",
    "name": "fieldName",
    "type": "select",
  },
  undefined,
]
`;

exports[`text defaults to text 1`] = `
Array [
  Object {
    "label": "the label",
    "name": "fieldName",
  },
  undefined,
]
`;

exports[`text passes initialValue 1`] = `
Array [
  Object {
    "label": "the label",
    "name": "fieldName",
    "type": "text",
  },
  "i am initial",
]
`;

exports[`text uses text 1`] = `
Array [
  Object {
    "label": "the label",
    "name": "fieldName",
    "type": "text",
  },
  undefined,
]
`;
